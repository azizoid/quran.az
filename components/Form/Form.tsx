import React, {
  ChangeEvent,
  SyntheticEvent,
  useContext,
  useEffect,
  useState,
} from 'react'
import { FormContext } from '../../store/form-store'
import soorah_list_object from '../../assets/soorah-list-object'
import translatorList from '../../assets/translatorList'

import { useRouter } from 'next/router'
import { FormProps } from '../../lib/types'
import { LoadingBoxes } from '../../ui/LoadingBoxes/LoadingBoxes'
import { getView } from '../../utility/getView/getView'

export const Form = (): JSX.Element => {
  const router = useRouter()
  const formContext = useContext(FormContext)

  const [state, setState] = useState<FormProps>()

  useEffect(() => setState(formContext), [formContext])

  const onHandleChange = (
    event: ChangeEvent<HTMLInputElement | HTMLSelectElement>
  ) => {
    const { name, value } = event.target

    switch (name) {
      case 'soorah':
        setState((prev) => ({
          ...prev,
          s: Number(value),
          a: '',
          q: '',
          view: name,
        }))
        break
      case 'ayah':
        setState((prev) => ({
          ...prev,
          s: prev.s,
          a: Number(value),
          q: '',
          view: name,
        }))
        break
      case 'search':
        setState((prev) => ({
          ...prev,
          s: 0,
          a: '',
          q: value,
          view: name,
        }))
        break
      case 'translator':
        setState((prev) => ({ ...prev, t: Number(value) }))
        break
    }
  }

  const onSubmit = (event: SyntheticEvent) => {
    event.preventDefault()

    const submitValue = getView(state)
    switch (submitValue.view) {
      case 'search':
        router.push(`/search/${state.q}?t=${state.t}`)
        break
      case 'soorah':
        router.push(`/${state.s}?t=${state.t}`)
        break
      case 'ayah':
        router.push(`/${state.s}/${state.a}?t=${state.t}`)
        break
      case 'empty':
      default:
        router.push(`/`)
    }
  }

  if (state?.view === 'init') {
    return <LoadingBoxes />
  }

  return (
    <form
      id="search"
      className="space-y-2 mb-6 alert alert-gray"
      acceptCharset="UTF-8"
      onSubmit={onSubmit}
    >
      <div className="form-row">
        <select
          className="
            form-control
            col-span-7
            focus:outline-none 
            focus:bg-white 
            focus:border-gray-500
            active:outline-none 
            active:border-gray-500
          "
          name="soorah"
          value={state?.s}
          onChange={onHandleChange}
        >
          <option value="0">Surələr:</option>
          {soorah_list_object.map(({ id, title }) => (
            <option value={id} key={id}>
              {title} surəsi. {id}
            </option>
          ))}
        </select>

        <input
          type="number"
          placeholder="Ayə"
          className="form-control col-span-2"
          name="ayah"
          size={3}
          maxLength={3}
          min={0}
          max={286}
          value={state?.a || ''}
          onChange={onHandleChange}
        />

        <select
          className="
            form-control
            col-span-3 text-center
            focus:outline-none 
            focus:bg-white 
            focus:border-gray-500
            active:outline-none 
            active:border-gray-500
          "
          name="translator"
          value={state?.t}
          onChange={onHandleChange}
        >
          {translatorList.map((soorah, index) => (
            <option value={index} key={index}>
              {soorah}
            </option>
          ))}
        </select>
      </div>

      <div className="form-row">
        <input
          type="text"
          placeholder="Kəlmə"
          className="form-control col-span-7"
          name="search"
          value={state?.q || ''}
          onChange={onHandleChange}
        />

        <button className="btn btn-success col-span-5" type="submit">
          Axtar
        </button>
      </div>
    </form>
  )
}

export default Form
