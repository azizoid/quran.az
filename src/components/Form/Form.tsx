import { ChangeEvent, SyntheticEvent, useContext, useEffect, useState } from 'react'

import { useRouter } from 'next/router'

import { soorahList } from '@/assets/soorah-list-object'
import { translatorList } from '@/assets/translatorList'
import { FormProps } from '@/lib/types'
import { getView } from '@/utility/getView/getView'

import { FormContext } from '../../store/form-store'

export const Form = (): JSX.Element => {
  const router = useRouter()
  const formContext = useContext(FormContext)

  const [state, setState] = useState<FormProps>(formContext)

  useEffect(() => {
    setState(formContext)
  }, [formContext])

  const onHandleChange = (event: ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
    const { name, value } = event.target

    switch (name) {
      case 'soorah':
        setState((prev) => ({
          ...prev,
          s: Number(value),
          a: '',
          q: '',
          view: name,
        }))
        break
      case 'ayah':
        setState((prev) => ({
          ...prev,
          s: prev.s,
          a: Number(value),
          q: '',
          view: name,
        }))
        break
      case 'search':
        setState((prev) => ({
          ...prev,
          s: 0,
          a: '',
          q: value,
          view: name,
        }))
        break
      case 'translator':
        setState((prev) => ({ ...prev, t: Number(value) }))
        break
    }
  }

  const onSubmit = (event: SyntheticEvent) => {
    event.preventDefault()

    const submitValue = getView(state)

    switch (submitValue.view) {
      case 'search':
        router.push(`/search/${submitValue.q}?t=${submitValue.t}`)
        break
      case 'soorah':
        router.push(`/${submitValue.s}?t=${submitValue.t}`)
        break
      case 'ayah':
        router.push(`/${submitValue.s}/${submitValue.a}?t=${submitValue.t}`)
        break
      case 'empty':
      default:
        router.push('/')
    }
  }

  return (
    <form
      id="search"
      className="space-y-2 mb-6 alert alert-gray"
      acceptCharset="UTF-8"
      onSubmit={onSubmit}
    >
      <div className="form-row">
        <select
          className="form-control col-span-7 focus:outline-none focus:bg-white focus:border-gray-500 active:outline-none active:border-gray-500"
          name="soorah"
          value={state?.s}
          onChange={onHandleChange}
        >
          <option value="0">Surələr:</option>
          {soorahList.map(({ id, title }) => (
            <option value={id} key={id}>
              {`${title} surəsi. ${id}`}
            </option>
          ))}
        </select>

        <input
          type="number"
          placeholder="Ayə"
          className="form-control col-span-2"
          name="ayah"
          size={3}
          maxLength={3}
          min={0}
          max={286}
          value={state?.a || ''}
          onChange={onHandleChange}
        />

        <select
          className="form-control col-span-3 text-center focus:outline-none focus:bg-white focus:border-gray-500 active:outline-none active:border-gray-500"
          name="translator"
          value={state?.t}
          onChange={onHandleChange}
        >
          {translatorList.map((soorah, index) => (
            <option value={index + 1} key={index}>
              {soorah}
            </option>
          ))}
        </select>
      </div>

      <div className="form-row">
        <input
          type="text"
          placeholder="Kəlmə"
          className="form-control col-span-7"
          name="search"
          value={state?.q || ''}
          onChange={onHandleChange}
        />

        <button className="btn btn-success col-span-5" type="submit">
          Axtar
        </button>
      </div>
    </form>
  )
}
