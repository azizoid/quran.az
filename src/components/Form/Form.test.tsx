import { render, fireEvent, screen } from '@testing-library/react'

import { Form } from './Form'

const defaultTranslator = process.env.NEXT_PUBLIC_DEFAULT_TRANSLATOR

const push = jest.fn()

jest.mock('next/navigation', () => ({
  useParams: () => ({}),
  useRouter: () => ({ push: push }),
  useSearchParams: () => ({
    get: jest.fn().mockReturnValue({ t: defaultTranslator }),
  }),
}))

jest.mock('@/ui/LoadingBoxes/LoadingBoxes', () => <span>Loader</span>)

describe('<Form>', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('Form Snapshot', () => {
    const { container } = render(<Form />)

    expect(screen.getAllByRole('combobox')).toHaveLength(2)
    expect(screen.getByRole('spinbutton')).toBeInTheDocument()
    expect(screen.getByRole('textbox')).toBeInTheDocument()
    expect(screen.getByRole('button', { name: 'Axtar' })).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  test('Click submit button on empty form', () => {
    render(<Form />)

    const submitBtn = screen.getByRole('button')
    fireEvent.click(submitBtn)

    expect(push).toHaveBeenLastCalledWith('/')
  })

  test('Change Soorah to 2 and click Submit Button', () => {
    render(<Form />)

    const submitBtn = screen.getByRole('button')
    const [soorahSelectBox] = screen.getAllByRole('combobox')

    fireEvent.change(soorahSelectBox, { target: { value: 2 } })
    expect(soorahSelectBox).toHaveValue('2')

    fireEvent.click(submitBtn)

    expect(push).toHaveBeenLastCalledWith(`/2?t=${defaultTranslator}`)
  })

  test('Select Soorah 2 and Ayah 23 and click the button', () => {
    render(<Form />)

    const submitBtn = screen.getByRole('button')
    const [soorahSelectBox] = screen.getAllByRole('combobox')
    const ayah = screen.getByRole('spinbutton')

    fireEvent.change(soorahSelectBox, { target: { value: 2 } })
    fireEvent.change(ayah, { target: { value: 23 } })

    fireEvent.click(submitBtn)
    expect(push).toHaveBeenLastCalledWith(`/2/23?t=${defaultTranslator}`)
  })

  test('Enter Search Query and click the button', () => {
    render(<Form />)

    const submitBtn = screen.getByRole('button')
    const query = screen.getByPlaceholderText('Kəlmə')

    fireEvent.change(query, { target: { value: 'Musa' } })

    fireEvent.click(submitBtn)
    expect(push).toHaveBeenLastCalledWith(`/search/Musa?t=${defaultTranslator}`)
  })

  test('Reset Soorah and Ayah on Query enter', () => {
    render(<Form />)

    const [soorahSelectBox, translatorSelectBox] = screen.getAllByRole('combobox')
    const ayah = screen.getByRole('spinbutton')
    const query = screen.getByPlaceholderText('Kəlmə')

    fireEvent.change(soorahSelectBox, { target: { value: 2 } })
    fireEvent.change(ayah, { target: { value: 23 } })

    expect(soorahSelectBox).toHaveValue('2')
    expect(ayah).toHaveValue(23)
    expect(translatorSelectBox).toHaveValue('4')
    expect(query).toHaveValue('')

    fireEvent.change(query, { target: { value: 'Musa' } })
    fireEvent.change(translatorSelectBox, { target: { value: 1 } })

    expect(soorahSelectBox).toHaveValue('0')
    expect(ayah).toHaveValue(null)
    expect(translatorSelectBox).toHaveValue('1')
    expect(query).toHaveValue('Musa')
  })
})
