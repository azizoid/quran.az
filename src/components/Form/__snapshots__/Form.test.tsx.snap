// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form> Form Snapshot 1`] = `
<div>
  <form
    accept-charset="UTF-8"
    class="space-y-2 mb-6 alert alert-gray"
    id="search"
  >
    <div
      class="form-row"
    >
      <select
        class="form-control col-span-7 focus:outline-none focus:bg-white focus:border-gray-500 active:outline-none active:border-gray-500"
        name="soorah"
      >
        <option
          value="0"
        >
          Surələr:
        </option>
        <option
          value="1"
        >
          Fatihə surəsi. 1
        </option>
        <option
          value="2"
        >
          Bəqərə surəsi. 2
        </option>
        <option
          value="3"
        >
          Ali İmran surəsi. 3
        </option>
        <option
          value="4"
        >
          Nisa surəsi. 4
        </option>
        <option
          value="5"
        >
          Maidə surəsi. 5
        </option>
        <option
          value="6"
        >
          Ənam surəsi. 6
        </option>
        <option
          value="7"
        >
          Əraf surəsi. 7
        </option>
        <option
          value="8"
        >
          Ənfal surəsi. 8
        </option>
        <option
          value="9"
        >
          Tovbə surəsi. 9
        </option>
        <option
          value="10"
        >
          Yunus surəsi. 10
        </option>
        <option
          value="11"
        >
          Hud surəsi. 11
        </option>
        <option
          value="12"
        >
          Yusuf surəsi. 12
        </option>
        <option
          value="13"
        >
          Rad surəsi. 13
        </option>
        <option
          value="14"
        >
          İbrahim surəsi. 14
        </option>
        <option
          value="15"
        >
          Hicr surəsi. 15
        </option>
        <option
          value="16"
        >
          Nəhl surəsi. 16
        </option>
        <option
          value="17"
        >
          İsra surəsi. 17
        </option>
        <option
          value="18"
        >
          Kəhf surəsi. 18
        </option>
        <option
          value="19"
        >
          Məryəm surəsi. 19
        </option>
        <option
          value="20"
        >
          Ta ha surəsi. 20
        </option>
        <option
          value="21"
        >
          Ənbiya surəsi. 21
        </option>
        <option
          value="22"
        >
          Həcc surəsi. 22
        </option>
        <option
          value="23"
        >
          Muminun surəsi. 23
        </option>
        <option
          value="24"
        >
          Nur surəsi. 24
        </option>
        <option
          value="25"
        >
          Furqan surəsi. 25
        </option>
        <option
          value="26"
        >
          Şuəra surəsi. 26
        </option>
        <option
          value="27"
        >
          Nəml surəsi. 27
        </option>
        <option
          value="28"
        >
          Qasas surəsi. 28
        </option>
        <option
          value="29"
        >
          Ənkəbut surəsi. 29
        </option>
        <option
          value="30"
        >
          Rum surəsi. 30
        </option>
        <option
          value="31"
        >
          Loğman surəsi. 31
        </option>
        <option
          value="32"
        >
          Səcdə surəsi. 32
        </option>
        <option
          value="33"
        >
          Əhzab surəsi. 33
        </option>
        <option
          value="34"
        >
          Səba surəsi. 34
        </option>
        <option
          value="35"
        >
          Fatir surəsi. 35
        </option>
        <option
          value="36"
        >
          Ya sin surəsi. 36
        </option>
        <option
          value="37"
        >
          Saffat surəsi. 37
        </option>
        <option
          value="38"
        >
          Sad surəsi. 38
        </option>
        <option
          value="39"
        >
          Zumər surəsi. 39
        </option>
        <option
          value="40"
        >
          Ğafir surəsi. 40
        </option>
        <option
          value="41"
        >
          Fussilət surəsi. 41
        </option>
        <option
          value="42"
        >
          Şura surəsi. 42
        </option>
        <option
          value="43"
        >
          Zuxruf surəsi. 43
        </option>
        <option
          value="44"
        >
          Duxan surəsi. 44
        </option>
        <option
          value="45"
        >
          Casiyə surəsi. 45
        </option>
        <option
          value="46"
        >
          Əhqaf surəsi. 46
        </option>
        <option
          value="47"
        >
          Muhəmməd surəsi. 47
        </option>
        <option
          value="48"
        >
          Fəth surəsi. 48
        </option>
        <option
          value="49"
        >
          Hucurat surəsi. 49
        </option>
        <option
          value="50"
        >
          Qaf surəsi. 50
        </option>
        <option
          value="51"
        >
          Zəriyət surəsi. 51
        </option>
        <option
          value="52"
        >
          Tur surəsi. 52
        </option>
        <option
          value="53"
        >
          Nəcm surəsi. 53
        </option>
        <option
          value="54"
        >
          Qamər surəsi. 54
        </option>
        <option
          value="55"
        >
          Rahmən surəsi. 55
        </option>
        <option
          value="56"
        >
          Vaqiə surəsi. 56
        </option>
        <option
          value="57"
        >
          Hədid surəsi. 57
        </option>
        <option
          value="58"
        >
          Mucadilə surəsi. 58
        </option>
        <option
          value="59"
        >
          Həşr surəsi. 59
        </option>
        <option
          value="60"
        >
          Mumtəhənə surəsi. 60
        </option>
        <option
          value="61"
        >
          Saff surəsi. 61
        </option>
        <option
          value="62"
        >
          Cumuə surəsi. 62
        </option>
        <option
          value="63"
        >
          Munafiqun surəsi. 63
        </option>
        <option
          value="64"
        >
          Təğabun surəsi. 64
        </option>
        <option
          value="65"
        >
          Talaq surəsi. 65
        </option>
        <option
          value="66"
        >
          Təhrim surəsi. 66
        </option>
        <option
          value="67"
        >
          Mulk surəsi. 67
        </option>
        <option
          value="68"
        >
          Qaləm surəsi. 68
        </option>
        <option
          value="69"
        >
          Haqqə surəsi. 69
        </option>
        <option
          value="70"
        >
          Məaric surəsi. 70
        </option>
        <option
          value="71"
        >
          Nuh surəsi. 71
        </option>
        <option
          value="72"
        >
          Cinn surəsi. 72
        </option>
        <option
          value="73"
        >
          Muzzəmmil surəsi. 73
        </option>
        <option
          value="74"
        >
          Muddəssir surəsi. 74
        </option>
        <option
          value="75"
        >
          Qiyamə surəsi. 75
        </option>
        <option
          value="76"
        >
          İnsan surəsi. 76
        </option>
        <option
          value="77"
        >
          Mursəlat surəsi. 77
        </option>
        <option
          value="78"
        >
          Nəbə surəsi. 78
        </option>
        <option
          value="79"
        >
          Naziat surəsi. 79
        </option>
        <option
          value="80"
        >
          Əbəsə surəsi. 80
        </option>
        <option
          value="81"
        >
          Təkvir surəsi. 81
        </option>
        <option
          value="82"
        >
          İnfitar surəsi. 82
        </option>
        <option
          value="83"
        >
          Mutaffifin surəsi. 83
        </option>
        <option
          value="84"
        >
          İnşiqaq surəsi. 84
        </option>
        <option
          value="85"
        >
          Buruc surəsi. 85
        </option>
        <option
          value="86"
        >
          Tariq surəsi. 86
        </option>
        <option
          value="87"
        >
          Əla surəsi. 87
        </option>
        <option
          value="88"
        >
          Ğaşiyə surəsi. 88
        </option>
        <option
          value="89"
        >
          Fəcr surəsi. 89
        </option>
        <option
          value="90"
        >
          Bələd surəsi. 90
        </option>
        <option
          value="91"
        >
          Şəms surəsi. 91
        </option>
        <option
          value="92"
        >
          Leyl surəsi. 92
        </option>
        <option
          value="93"
        >
          Duha surəsi. 93
        </option>
        <option
          value="94"
        >
          Şərh surəsi. 94
        </option>
        <option
          value="95"
        >
          Tin surəsi. 95
        </option>
        <option
          value="96"
        >
          Ələq surəsi. 96
        </option>
        <option
          value="97"
        >
          Qədr surəsi. 97
        </option>
        <option
          value="98"
        >
          Beyyinə surəsi. 98
        </option>
        <option
          value="99"
        >
          Zəlzələ surəsi. 99
        </option>
        <option
          value="100"
        >
          Adiyat surəsi. 100
        </option>
        <option
          value="101"
        >
          Qariə surəsi. 101
        </option>
        <option
          value="102"
        >
          Təkasur surəsi. 102
        </option>
        <option
          value="103"
        >
          Əsr surəsi. 103
        </option>
        <option
          value="104"
        >
          Huməzə surəsi. 104
        </option>
        <option
          value="105"
        >
          Fil surəsi. 105
        </option>
        <option
          value="106"
        >
          Qureyş surəsi. 106
        </option>
        <option
          value="107"
        >
          Maun surəsi. 107
        </option>
        <option
          value="108"
        >
          Kovsər surəsi. 108
        </option>
        <option
          value="109"
        >
          Kafirun surəsi. 109
        </option>
        <option
          value="110"
        >
          Nəsr surəsi. 110
        </option>
        <option
          value="111"
        >
          Məsəd surəsi. 111
        </option>
        <option
          value="112"
        >
          İxlas surəsi. 112
        </option>
        <option
          value="113"
        >
          Fələq surəsi. 113
        </option>
        <option
          value="114"
        >
          Nas surəsi. 114
        </option>
      </select>
      <input
        class="form-control col-span-2"
        max="286"
        maxlength="3"
        min="0"
        name="ayah"
        placeholder="Ayə"
        size="3"
        type="number"
        value=""
      />
      <select
        class="form-control col-span-3 text-center focus:outline-none focus:bg-white focus:border-gray-500 active:outline-none active:border-gray-500"
        name="translator"
      >
        <option
          value="1"
        >
          Əlixan Musayev
        </option>
        <option
          value="2"
        >
          Bünyadov-Məmmədəliyev
        </option>
        <option
          value="3"
        >
          Эльмир Кулиев
        </option>
        <option
          value="4"
        >
          Ələddin Sultanov
        </option>
      </select>
    </div>
    <div
      class="form-row"
    >
      <input
        class="form-control col-span-7"
        name="search"
        placeholder="Kəlmə"
        type="text"
        value=""
      />
      <button
        class="btn btn-success col-span-5"
        type="submit"
      >
        Axtar
      </button>
    </div>
  </form>
</div>
`;
